skcosmo/sparse/IncrementalSparseKRR.py

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python
+
+import os
+import sys
+import numpy as np
+from tools import sorted_eigh
+
+class IncrementalSparseKRR(object):
+    """
+        Performs sparsified kernel ridge regression. Example usage:
+
+        KMM = build_kernel(Xm, Xm)
+        iskrr = IncrementalSparseKRR()
+        iskrr.initialize_fit(KMM)
+        for i in batches:
+            KNMi = build_kernel(Xi, Xm)
+            iskrr.fit_batch(KNMi, Yi)
+        iskrr.finalize_fit()
+        for i in batches:
+            KNMi = build_kernel(Xi, Xm)
+            iskrr.predict(KNMi)
+
+        ---Attributes---
+        sigma: regularization parameter
+        regularization: additional regularization
+        regularization_type: type of regularization.
+            Choices are 'scalar' (constant regularization),
+            or 'max_eig' (regularization based on maximum eigenvalue)
+        W: regression weights
+        KMM: centered kernel between representative points
+        KY: product of the KNM kernel and the properties Y
+        
+        ---Methods---
+        initialize_fit: initialize the fit of the sparse KRR
+            (i.e., store KMM)
+        fit_batch: fit a batch of training data
+        finalize_fit: finalize the KRR fitting
+            (i.e., compute regression weights)
+        predict: compute predicted Y values
+        
+        ---References---
+        1.  M. Ceriotti, M. J. Willatt, G. Csanyi,
+            'Machine Learning of Atomic-Scale Properties
+            Based on Physical Principles', Handbook of Materials Modeling,
+            Springer, 2018
+        2.  A. J. Smola, B. Scholkopf, 'Sparse Greedy Matrix Approximation 
+            for Machine Learning', Proceedings of the 17th International
+            Conference on Machine Learning, 911-918, 2000
+    """
+    
+    def __init__(self, sigma=1.0, regularization=1.0E-12, 
+            regularization_type='scalar', rcond=None):
+        self.sigma = sigma
+        self.regularization = regularization
+        self.regularization_type = regularization_type
+        self.rcond = rcond
+        self.W = None
+        self.KX = None
+        self.KY = None
+        self.y_dim = None
+
+    def initialize_fit(self, KMM, y_dim=1):
+        """
+            Initialize the KRR fitting by computing the
+            eigendecomposition of KMM
+
+            ---Arguments---
+            KMM: centered kernel between the representative points
+            y_dim: number of properties
+        """
+
+        # Check for valid Y dimension
+        if y_dim < 1:
+            print("Y dimension must be at least 1")
+            return
+
+        self.y_dim = y_dim
+        self.KX = self.sigma*KMM
```

Comment:
This is unusual for me, where does this kind of regularization come from? Or when is this useful?

Reply:
When doing these "sparse" methods based on subsets of the data points, this is where the "traditional" regularization comes in (we have a derivation in the [KPCovR paper](https://doi.org/10.1088/2632-2153/aba9ef) and there is one in [this publication](https://dl.acm.org/doi/10.5555/645529.657980) as well). The "sigma" variable name is just an adoption from the GPR literature. The "add a scalar times the identity" type of regularization that is also included below doesn't have a rigorous basis in this method, but I have found it helps in some cases.

Reply:
so following the publication of Smole and Schölkopf you ve sent, this term corresponds to Eq. 7 the m\Lambda KNN without m (which you probably get through the for loop), doesnt it?

Reply:
The _m_ doesn't appear from the for loop, instead it's just absorbed into the `self.sigma` since it's a constant.